src/legacy/core_plugins/kibana/public/visualize/embeddable/visualize_embeddable.ts

```diff
@@ -211,58 +261,152 @@ export class VisualizeEmbeddable extends Embeddable<VisualizeInput, VisualizeOut
    */
-  public render(domNode: HTMLElement) {
+  public async render(domNode: HTMLElement) {
     this.timeRange = _.cloneDeep(this.input.timeRange);
     this.query = this.input.query;
     this.filters = this.input.filters;
 
     this.transferCustomizationsToUiState();
 
-    const dataAttrs: { [key: string]: string } = {
-      'shared-item': '',
-      title: this.output.title || '',
+    this.savedVisualization.vis._setUiState(this.uiState);
+    this.uiState = this.savedVisualization.vis.getUiState();
+
+    // init default actions
+    forEach(this.vis.type.events, (event, eventName) => {
+      if (event.disabled || !eventName) {
+        return;
+      } else {
+        this.actions[eventName] = event.defaultAction;
+      }
+    });
+
+    // This is a hack to give maps visualizations access to data in the
+    // globalState, since they can no longer access it via searchSource.
+    // TODO: Remove this as a part of elastic/kibana#30593
+    this.vis.API.getGeohashBounds = () => {
+      return queryGeohashBounds(this.savedVisualization.vis, {
+        filters: this.filters,
+        query: this.query,
+        searchSource: this.savedVisualization.searchSource,
+      });
+    };
+
+    // this is a hack to make editor still work, will be removed once we clean up editor
+    this.vis.hasInspector = () => {
+      if (['markdown', 'input_control_vis', 'metrics', 'vega'].includes(this.vis.type.name)) {
+        return false;
+      }
+      return this.getInspectorAdapters();
+    };
+
+    this.vis.openInspector = this.openInspector;
+
+    this.vis.filters = { timeRange: this.timeRange };
+
+    const expressionParams: IExpressionLoaderParams = {
+      searchContext: {
+        filters: this.filters,
+        timeRange: this.timeRange,
+        query: this.query,
+        type: 'kibana_context',
+      },
+      extraHandlers: {
+        vis: this.vis,
+        uiState: this.uiState,
+      },
     };
+
+    const div = document.createElement('div');
+    div.className = `visualize panel-content panel-content--fullWidth`;
+    domNode.appendChild(div);
+    this.domNode = div;
+
+    this.expression = await buildPipeline(this.vis, {
+      searchSource: this.savedVisualization.searchSource,
+      timeRange: this.timeRange,
+    });
+    if (!this.handler) {
+      this.handler = new expressions.ExpressionLoader(div, this.expression, expressionParams);
+    } else {
+      this.handler.update(this.expression, expressionParams);
+    }
+
+    this.subscriptions.push(
+      this.handler.events$.subscribe(async event => {
+        if (this.actions[event.name]) {
+          event.data.aggConfigs = getTableAggs(this.vis);
+          const filters: esFilters.Filter[] = this.actions[event.name](event.data) || [];
+          const mappedFilters = filters.map(mapFilter);
+          const timeFieldName = this.vis.indexPattern.timeFieldName;
+
+          npStart.plugins.uiActions.executeTriggerActions(APPLY_FILTER_TRIGGER, {
```

Comment:
Tested apply filters popover works in both visualization and dashboard :heavy_check_mark: 

Since apply filters popover now uses the React component directly, you can delete the following code, to avoid confusion (tested on own machine - and it works):

 - `apply-filters-popover` usage in `dashboard_app.html`
 - `applyFiltersPopover` and `applyFiltersPopoverHelper` directives in `src/legacy/core_plugins/data/public/shim/legacy_module.ts`
 - `onCancelApplyFilters`, `onApplyFilters`  and `$scope.$watch('appState.$newFilters'` from `dashboard_app_controller.tsx`.